Pydanticgen - Modularize template

sneakers-the-rat opened the pull request:
Another cleanup step for the pydantic generator - modularizing the templates and coupling them with models. Typing quickly and opening a draft rn but my battery is about to die and so i'll edit this with more info once i find my charger.

pretty big PR but this one makes very few changes to the actual behavior of the generator. mostly a refactoring

goals:
- be able to generate each of the components of a schema separately for testing and also for runtime use
- be able to inspect what's going on in the template more easily
- have models for the endpoint so it is clear what the generator supports - eg. the attribute model supports `maximum_value` etc. but does not support literally all possible values for a slot definition. 
- make it clearer where one should hook into the generation process
- make future steps of cleaning up the pydantic generator easier by having less anonymous objects

todo:
- [x] clean up the code and templates - i have a hell of a time with jinja whitespace formatting so i just unindented everything, but it could be made more readable. Also there are some odd ordering and mix and match syntax i need to fix
- [x] make tests pass - i thought i had replicated the existing template pretty exactly, but i'm getting mysterious errors from inlining and unenforced constraints. will follow up tmrw
- [x] investigate kwargs unpacking in jinja templates because it's pretty brittle to need to hardcode all the arguments in the macro calls - not possible
- [x] test new functionality - imports `__add__` method most notably, rest is pretty testing neutral so far.

edit: ok returned

## template structure 

the strategy is pretty self explanatory i think, split each separate part of the template into its own sub-template. i used jinja macros here to keep the iteration in the top level template so each individual template could be used on its own. As is the top level `PydanticModule` class holds everything that should be given to the `module` template, and that holds for the rest of the subtemplates except for `attribute` and `validator`. 

### alternative

Another strategy I was thinking about was to give the `TemplateModel` metaclass a `build` method that subclasses could override that does the actual rendering for each piece, and then they get stitched together - that would make for nice clean build syntax where the relationship between the model and the template is super clear, but wouldn't be the single jinja template render call, idk what's good there.

## generator changes

### template values

i have thought it odd that there are two additional template properties for precomputed field values and class bases, so i moved both of those to the attribute and class models, respectively. 

I also changed the pydantic version parameter to an int because it is so it would be possible to make range comparisons in case godforbid there is a v3 debacle. at the moment i'm seeing some places where it's still a string and will fix. this change was made in this PR because pydantic version has to propagate down through the templates and i was getting iffy with it being a string

### imports

sorry to change the imports structure so soon, i mentioned i didn't really like the implicit nature of the import syntax i had drafted, and this is more explicit, and thinking about usage i don't think authorship friendliness is as important as it is in the metamodel so more verbose models that can be more easily yielded from intermediate build stages is imo preferable to needing to do a bunch of `isinstance` checks all over the place.

i went ahead and made a little `Imports` collection class to make adding imports possible, resolving duplicates, but the implementation is pretty janky and needs to be tested. I think that will be nice to make it so when we linearize the build system then we can have each stage return "this class needs these things to be imported" and not have to worry about whether we already have those in the template or not. That should eventually make it so we don't need to declare top-level default imports at all, but instead dynamically generate the minimal necessary imports for a given schema, helping modularity in the long-term where any given feature/plugin can have its own build stage that spits out dependencies, eg. with arrays

Here's an example of the imports model, the `Imports.__add__` method (actually the `Import.merge` method) handles deduplicating and merging imports from the same module. 

```python
DEFAULT_IMPORTS = (
    Imports()
    + Import(module="__future__", objects=[ObjectImport(name="annotations")])
    + Import(module="datetime", objects=[ObjectImport(name="datetime"), ObjectImport(name="date")])
    + Import(module="re")
    + Import(
        module="typing",
        objects=[
            ObjectImport(name="Any"),
            ObjectImport(name="List"),
            ObjectImport(name="Literal"),
            ObjectImport(name="Dict"),
            ObjectImport(name="Optional"),
            ObjectImport(name="Union"),
        ],
    )
    + Import(module="pydantic.version", objects=[ObjectImport(name="VERSION", alias="PYDANTIC_VERSION")])
    + ConditionalImport(
        condition="int(PYDANTIC_VERSION[0])>=2",
        module="pydantic",
        objects=[
            ObjectImport(name="BaseModel"),
            ObjectImport(name="ConfigDict"),
            ObjectImport(name="Field"),
            ObjectImport(name="field_validator"),
        ],
        alternative=Import(
            module="pydantic",
            objects=[ObjectImport(name="BaseModel"), ObjectImport(name="Field"), ObjectImport(name="validator")],
        ),
    )
)
```

makes:

```python
from __future__ import annotations 
from datetime import (
    datetime,
    date
)

import re

from typing import (
    Any,
    List,
    Literal,
    Dict,
    Optional,
    Union
)

from pydantic.version import VERSION  as PYDANTIC_VERSION 
if int(PYDANTIC_VERSION[0])>=2:
    from pydantic import (
        BaseModel,
        ConfigDict,
        Field,
        field_validator
    )

else:
    from pydantic import (
        BaseModel,
        Field,
        validator
    )
```

### temporary mess

pending a reworking of the serialization method, there is a bit of mess at the end where i'm shoehorning the data into the new models before passing to the template. that's super undesirable and i'm taking a look now how else i could clean that up, but I think what would be worth doing next is to tidy that method up generally and that might fit better in another PR. 

## next steps

with models in hand, we should use them in the generator - currently there is a lot of implicit mutation done throughout this and the OO generator, and so we could start linearizing that out to do one pass through to make classes, one pass through to fix their attributes, etc. I think ultimately it is desirable when translating between formats like this to have source and target models - i like the idea of just using the metamodel dataclasses everywhere, but it ends up having a few not fun qualities:
- you end up needing to mutate the source models a lot, which can make things opaque pretty quickly, and also gets you into places like this https://github.com/linkml/linkml-runtime/pull/296 where you need to copy your way through things. 
- you also end up abusing the semantics of the source models, stuffing things in where they can fit rather than where they make sense
- it makes it actually a bit harder to do multiple translations to related formats - the OOgenerator is a nice start, but since it's hard to know where the metamodel classes end, the oogenerator logic begins, and then on through to the templates, then i am left not really knowing what i should be reusing and trying to put in OOgenerator and what goes into pydanticgen. having models for the target domains lets you know what comes out of one leg of the translation process and lets you layer on complexity.
- as alluded to above, it's sort of hard to know what of the metamodel the generators support, so linking that to specific models that instantiate a given feature then 
- much easier testing :). the pydantic models validating are sort of a test on their own, but being able to break up the individual pieces and test a known endpoint model is good for the nerves <3

sneakers-the-rat commented:
Sample generation using kitchen sink at [acf2bdb](https://github.com/linkml/linkml/pull/1927/commits/acf2bdbe6c52e518ba0ade316a03994c2502c928)

## Pydantic 1

<details>
<summary>expand/collapse</summary>

```python
from __future__ import annotations 
from datetime import (
    datetime,
    date
)

from decimal import Decimal 
from enum import Enum 
import re

from typing import (
    Any,
    List,
    Literal,
    Dict,
    Optional,
    Union
)

from pydantic.version import VERSION  as PYDANTIC_VERSION 
if int(PYDANTIC_VERSION[0])>=2:
    from pydantic import (
        BaseModel,
        ConfigDict,
        Field,
        field_validator
    )

else:
    from pydantic import (
        BaseModel,
        Field,
        validator
    )


metamodel_version = "None"
version = "None"


class WeakRefShimBaseModel(BaseModel):
    __slots__ = '__weakref__'

class ConfiguredBaseModel(WeakRefShimBaseModel,
                validate_assignment = True,
                validate_all = True,
                underscore_attrs_are_private = True,
                extra = "forbid",
                arbitrary_types_allowed = True,
                use_enum_values = True):
    pass


class FamilialRelationshipType(str, Enum):
    SIBLING_OF = "SIBLING_OF"
    PARENT_OF = "PARENT_OF"
    CHILD_OF = "CHILD_OF"


class DiagnosisType(str, Enum):
    TODO = "TODO"


class EmploymentEventType(str, Enum):
    """
    codes for different kinds of employment/HR related events
    """
    # event for a new employee
    HIRE = "HIRE"
    FIRE = "FIRE"
    # promotion event
    PROMOTION = "PROMOTION"
    # transfer internally
    TRANSFER = "TRANSFER"


class OtherCodes(str, Enum):
    a_b = "a b"


class LifeStatusEnum(str, Enum):
    LIVING = "LIVING"
    DEAD = "DEAD"
    UNKNOWN = "UNKNOWN"


class CordialnessEnum(str, Enum):
    # warm and hearty friendliness
    heartfelt = "heartfelt"
    # spiteful
    hateful = "hateful"
    # not overly friendly nor obnoxiously spiteful
    indifferent = "indifferent"


class Activity(ConfiguredBaseModel):
    """
    a provence-generating activity
    """
    id: str = Field(...)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    was_informed_by: Optional[str] = Field(None)
    was_associated_with: Optional[str] = Field(None)
    used: Optional[str] = Field(None)
    description: Optional[str] = Field(None)


class Agent(ConfiguredBaseModel):
    """
    a provence-generating agent
    """
    id: str = Field(...)
    acted_on_behalf_of: Optional[str] = Field(None)
    was_informed_by: Optional[str] = Field(None)


class HasAliases(ConfiguredBaseModel):
    aliases: Optional[List[str]] = Field(default_factory=list)


class Friend(ConfiguredBaseModel):
    name: Optional[str] = Field(None)


class Person(HasAliases):
    """
    A person, living or dead
    """
    id: str = Field(...)
    name: Optional[str] = Field(None)
    has_employment_history: Optional[List[EmploymentEvent]] = Field(None)
    has_familial_relationships: Optional[List[FamilialRelationship]] = Field(None)
    has_medical_history: Optional[List[MedicalEvent]] = Field(None)
    age_in_years: Optional[int] = Field(None, description="""number of years since birth""", ge=0, le=999)
    addresses: Optional[List[Address]] = Field(default_factory=list)
    has_birth_event: Optional[BirthEvent] = Field(None)
    species_name: Optional[str] = Field(None)
    stomach_count: Optional[int] = Field(None    , le=1, ge=1)
    is_living: Optional[LifeStatusEnum] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)

    @validator('name', allow_reuse=True)
    def pattern_name(cls, v):
        pattern=re.compile(r"^\S+ \S+$")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid name format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid name format: {v}")
        return v


    @validator('species_name', allow_reuse=True)
    def pattern_species_name(cls, v):
        pattern=re.compile(r"^[A-Z]+[a-z]+(-[A-Z]+[a-z]+)?\\.[A-Z]+(-[0-9]{4})?$")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid species_name format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid species_name format: {v}")
        return v



class Organization(HasAliases):
    """
    An organization.

    This description
    includes newlines

    ## Markdown headers

     * and
     * a
     * list
    """
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class Place(HasAliases):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class Address(ConfiguredBaseModel):
    street: Optional[str] = Field(None)
    city: Optional[str] = Field(None)
    altitude: Optional[Decimal] = Field(None)


class Concept(ConfiguredBaseModel):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class DiagnosisConcept(Concept):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class ProcedureConcept(Concept):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class Event(ConfiguredBaseModel):
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class Relationship(ConfiguredBaseModel):
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    related_to: Optional[str] = Field(None)
    type: Optional[str] = Field(None)
    cordialness: Optional[CordialnessEnum] = Field(None)


class FamilialRelationship(Relationship):
    cordialness: Optional[CordialnessEnum] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    related_to: str = Field(...)
    type: FamilialRelationshipType = Field(...)


class BirthEvent(Event):
    in_location: Optional[str] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class EmploymentEvent(Event):
    employed_at: Optional[str] = Field(None)
    type: Optional[EmploymentEventType] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class MedicalEvent(Event):
    in_location: Optional[str] = Field(None)
    diagnosis: Optional[DiagnosisConcept] = Field(None)
    procedure: Optional[ProcedureConcept] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class WithLocation(ConfiguredBaseModel):
    in_location: Optional[str] = Field(None)


class MarriageEvent(WithLocation, Event):
    married_to: Optional[str] = Field(None)
    in_location: Optional[str] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class Company(Organization):
    ceo: Optional[str] = Field(None)
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class CodeSystem(ConfiguredBaseModel):
    id: str = Field(...)
    name: Optional[str] = Field(None)


class Dataset(ConfiguredBaseModel):
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")
    persons: Optional[List[Person]] = Field(default_factory=list)
    companies: Optional[List[Company]] = Field(default_factory=list)
    activities: Optional[List[Activity]] = Field(default_factory=list)
    code_systems: Optional[List[CodeSystem]] = Field(None)


class FakeClass(ConfiguredBaseModel):
    test_attribute: Optional[str] = Field(None)


class ClassWithSpaces(ConfiguredBaseModel):
    slot_with_space_1: Optional[str] = Field(None)


class SubclassTest(ClassWithSpaces):
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


class SubSubClass2(SubclassTest):
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


class TubSubClass1(SubclassTest):
    """
    Same depth as Sub sub class 1
    """
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


# Update forward refs
# see https://pydantic-docs.helpmanual.io/usage/postponed_annotations/
Activity.update_forward_refs()
Agent.update_forward_refs()
HasAliases.update_forward_refs()
Friend.update_forward_refs()
Person.update_forward_refs()
Organization.update_forward_refs()
Place.update_forward_refs()
Address.update_forward_refs()
Concept.update_forward_refs()
DiagnosisConcept.update_forward_refs()
ProcedureConcept.update_forward_refs()
Event.update_forward_refs()
Relationship.update_forward_refs()
FamilialRelationship.update_forward_refs()
BirthEvent.update_forward_refs()
EmploymentEvent.update_forward_refs()
MedicalEvent.update_forward_refs()
WithLocation.update_forward_refs()
MarriageEvent.update_forward_refs()
Company.update_forward_refs()
CodeSystem.update_forward_refs()
Dataset.update_forward_refs()
FakeClass.update_forward_refs()
ClassWithSpaces.update_forward_refs()
SubclassTest.update_forward_refs()
SubSubClass2.update_forward_refs()
TubSubClass1.update_forward_refs()

```

</details>

## Pydantic 2

<details>
<summary>expand/collapse</summary>

```python
from __future__ import annotations 
from datetime import (
    datetime,
    date
)

from decimal import Decimal 
from enum import Enum 
import re

from typing import (
    Any,
    List,
    Literal,
    Dict,
    Optional,
    Union
)

from pydantic.version import VERSION  as PYDANTIC_VERSION 
if int(PYDANTIC_VERSION[0])>=2:
    from pydantic import (
        BaseModel,
        ConfigDict,
        Field,
        field_validator
    )

else:
    from pydantic import (
        BaseModel,
        Field,
        validator
    )


metamodel_version = "None"
version = "None"


class ConfiguredBaseModel(BaseModel):
    model_config = ConfigDict(
        validate_assignment = True,
        validate_default = True,
        extra = "forbid",
        arbitrary_types_allowed = True,
        use_enum_values = True)
    pass


class FamilialRelationshipType(str, Enum):
    SIBLING_OF = "SIBLING_OF"
    PARENT_OF = "PARENT_OF"
    CHILD_OF = "CHILD_OF"


class DiagnosisType(str, Enum):
    TODO = "TODO"


class EmploymentEventType(str, Enum):
    """
    codes for different kinds of employment/HR related events
    """
    # event for a new employee
    HIRE = "HIRE"
    FIRE = "FIRE"
    # promotion event
    PROMOTION = "PROMOTION"
    # transfer internally
    TRANSFER = "TRANSFER"


class OtherCodes(str, Enum):
    a_b = "a b"


class LifeStatusEnum(str, Enum):
    LIVING = "LIVING"
    DEAD = "DEAD"
    UNKNOWN = "UNKNOWN"


class CordialnessEnum(str, Enum):
    # warm and hearty friendliness
    heartfelt = "heartfelt"
    # spiteful
    hateful = "hateful"
    # not overly friendly nor obnoxiously spiteful
    indifferent = "indifferent"


class Activity(ConfiguredBaseModel):
    """
    a provence-generating activity
    """
    id: str = Field(...)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    was_informed_by: Optional[str] = Field(None)
    was_associated_with: Optional[str] = Field(None)
    used: Optional[str] = Field(None)
    description: Optional[str] = Field(None)


class Agent(ConfiguredBaseModel):
    """
    a provence-generating agent
    """
    id: str = Field(...)
    acted_on_behalf_of: Optional[str] = Field(None)
    was_informed_by: Optional[str] = Field(None)


class HasAliases(ConfiguredBaseModel):
    aliases: Optional[List[str]] = Field(default_factory=list)


class Friend(ConfiguredBaseModel):
    name: Optional[str] = Field(None)


class Person(HasAliases):
    """
    A person, living or dead
    """
    id: str = Field(...)
    name: Optional[str] = Field(None)
    has_employment_history: Optional[List[EmploymentEvent]] = Field(None)
    has_familial_relationships: Optional[List[FamilialRelationship]] = Field(None)
    has_medical_history: Optional[List[MedicalEvent]] = Field(None)
    age_in_years: Optional[int] = Field(None, description="""number of years since birth""", ge=0, le=999)
    addresses: Optional[List[Address]] = Field(default_factory=list)
    has_birth_event: Optional[BirthEvent] = Field(None)
    species_name: Optional[str] = Field(None)
    stomach_count: Optional[int] = Field(None    , le=1, ge=1)
    is_living: Optional[LifeStatusEnum] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)

    @field_validator('name')
    def pattern_name(cls, v):
        pattern=re.compile(r"^\S+ \S+$")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid name format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid name format: {v}")
        return v


    @field_validator('species_name')
    def pattern_species_name(cls, v):
        pattern=re.compile(r"^[A-Z]+[a-z]+(-[A-Z]+[a-z]+)?\\.[A-Z]+(-[0-9]{4})?$")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid species_name format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid species_name format: {v}")
        return v



class Organization(HasAliases):
    """
    An organization.

    This description
    includes newlines

    ## Markdown headers

     * and
     * a
     * list
    """
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class Place(HasAliases):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class Address(ConfiguredBaseModel):
    street: Optional[str] = Field(None)
    city: Optional[str] = Field(None)
    altitude: Optional[Decimal] = Field(None)


class Concept(ConfiguredBaseModel):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class DiagnosisConcept(Concept):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class ProcedureConcept(Concept):
    id: str = Field(...)
    name: Optional[str] = Field(None)
    in_code_system: Optional[str] = Field(None)


class Event(ConfiguredBaseModel):
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class Relationship(ConfiguredBaseModel):
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    related_to: Optional[str] = Field(None)
    type: Optional[str] = Field(None)
    cordialness: Optional[CordialnessEnum] = Field(None)


class FamilialRelationship(Relationship):
    cordialness: Optional[CordialnessEnum] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    related_to: str = Field(...)
    type: FamilialRelationshipType = Field(...)


class BirthEvent(Event):
    in_location: Optional[str] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class EmploymentEvent(Event):
    employed_at: Optional[str] = Field(None)
    type: Optional[EmploymentEventType] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class MedicalEvent(Event):
    in_location: Optional[str] = Field(None)
    diagnosis: Optional[DiagnosisConcept] = Field(None)
    procedure: Optional[ProcedureConcept] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class WithLocation(ConfiguredBaseModel):
    in_location: Optional[str] = Field(None)


class MarriageEvent(WithLocation, Event):
    married_to: Optional[str] = Field(None)
    in_location: Optional[str] = Field(None)
    started_at_time: Optional[date] = Field(None)
    ended_at_time: Optional[date] = Field(None)
    is_current: Optional[bool] = Field(None)
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")


class Company(Organization):
    ceo: Optional[str] = Field(None)
    id: str = Field(...)
    name: Optional[str] = Field(None)
    aliases: Optional[List[str]] = Field(default_factory=list)


class CodeSystem(ConfiguredBaseModel):
    id: str = Field(...)
    name: Optional[str] = Field(None)


class Dataset(ConfiguredBaseModel):
    metadata: Optional[Any] = Field(None, description="""Example of a slot that has an unconstrained range""")
    persons: Optional[List[Person]] = Field(default_factory=list)
    companies: Optional[List[Company]] = Field(default_factory=list)
    activities: Optional[List[Activity]] = Field(default_factory=list)
    code_systems: Optional[List[CodeSystem]] = Field(None)


class FakeClass(ConfiguredBaseModel):
    test_attribute: Optional[str] = Field(None)


class ClassWithSpaces(ConfiguredBaseModel):
    slot_with_space_1: Optional[str] = Field(None)


class SubclassTest(ClassWithSpaces):
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


class SubSubClass2(SubclassTest):
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


class TubSubClass1(SubclassTest):
    """
    Same depth as Sub sub class 1
    """
    slot_with_space_2: Optional[ClassWithSpaces] = Field(None)
    slot_with_space_1: Optional[str] = Field(None)


# Model rebuild
# see https://pydantic-docs.helpmanual.io/usage/models/#rebuilding-a-model
Activity.model_rebuild()
Agent.model_rebuild()
HasAliases.model_rebuild()
Friend.model_rebuild()
Person.model_rebuild()
Organization.model_rebuild()
Place.model_rebuild()
Address.model_rebuild()
Concept.model_rebuild()
DiagnosisConcept.model_rebuild()
ProcedureConcept.model_rebuild()
Event.model_rebuild()
Relationship.model_rebuild()
FamilialRelationship.model_rebuild()
BirthEvent.model_rebuild()
EmploymentEvent.model_rebuild()
MedicalEvent.model_rebuild()
WithLocation.model_rebuild()
MarriageEvent.model_rebuild()
Company.model_rebuild()
CodeSystem.model_rebuild()
Dataset.model_rebuild()
FakeClass.model_rebuild()
ClassWithSpaces.model_rebuild()
SubclassTest.model_rebuild()
SubSubClass2.model_rebuild()
TubSubClass1.model_rebuild()

```

</details>

sneakers-the-rat commented:
this is ready for review :) @pkalita-lbl @kevinschaper  (just tagging the 2 ppl who reviewed the last pydanticgen PR, sry if thats annoying)

sneakers-the-rat commented:
Can I get a rough nod on this so I can merge it into https://github.com/linkml/linkml/pull/1887 and keep working on it, and then merge any changes from this review into it later? Next steps there are to handle injecting imports and classes and then it'll be good to go, and I need the changes here to imports and templates to make that tidy.

Just looking for will consider this or not pending review changes, but no hurry on the actual review bc I have a decent amount of work to do there still as well with testing

cmungall commented:
I like the direction this is going in, the refactoring makes sense to me

kevinschaper reviewed (approved):
I love it. 

Something I considered but hadn't looked into was whether to just always pass the output of the template into black (or ruff now?), so that we don't have to get the jinja just right.  

Sort of related, I was looking at how https://github.com/apple/pkl is doing code generation, they use https://github.com/square/javapoet (and a kotlinpoet parallel repo). It might be worth looking for ideas in that codebase that are worth taking inspiration from. It doesn't use templating at all, I'm not sure what I think about that, I think for all the awkwardness of jinja, I feel like it gives a good road map to the whole process. 

The template model is also great for splitting up the testing, so that we can test schema -> template model and template model -> pydantic code separately and not have everything be quite so end-to-end.

sneakers-the-rat commented:
> Something I considered but hadn't looked into was whether to just always pass the output of the template into black (or ruff now?), so that we don't have to get the jinja just right.

omg i was about to do this so many times but i didn't want to add a runtime dependency, this would actually be super good to do bc i absolutely despise dealing with jinja whitespace. I read through the black source to see how one would use it to just "format this entire module as a string" but couldn't quite figure it out, but gave up pretty quickly, it would probably take 10m to do.

>  It might be worth looking for ideas in that codebase that are worth taking inspiration from. 

checking it out rn, but from a brief scan of [this](https://github.com/apple/pkl/blob/main/pkl-codegen-kotlin/src/main/kotlin/org/pkl/codegen/kotlin/KotlinCodeGenerator.kt) i am not necessarily flooded with envy, is there a codegen you were looking at in particular?

> It doesn't use templating at all, I'm not sure what I think about that, I think for all the awkwardness of jinja, I feel like it gives a good road map to the whole process.

Yeah I think it's largely a matter of taste. Templating engines like jinja sorta blur the line between template and programming language. I tend to think of it like the template just does layout from fixed values, and no computation except if switches should happen in the template (not an ironclad rules obvi). Templating is good because if it's all imperative then you can't really know what you expect, that's sorta the goal of this PR - make the values going into the template explicit, make the different sections of the template explicit, so you can easily find what values you need and how they'll be used. 

I mentioned this above, but i am 50/50 on whether giving each `TemplateModel` a render method that renders its specific subtemplate and they're just `''.join([])`ed at the end (or passed as strings to the `module` template), i think that i'm going to make a version that does that and see how it feels, because that makes the connection between the objects and the templates much more explicit, and would also allow us to not use `macro`s which are sorta awkward.

> The template model is also great for splitting up the testing, so that we can test schema -> template model and template model -> pydantic code separately and not have everything be quite so end-to-end.

yes exactly! and i'd also like to eventually make these target models be able to report what features of the metamodel the generator is capable of supporting - if the target has some means of rendering an array, handle things like min/max value ranges, etc. So then in compliance tests we can just confirm that the given generator meets its advertised features rather than have so many manual if/else statements in the tests which can be pretty brittle (i see a lot of commented out and skipped tests, which is understandable)

sneakers-the-rat commented:
actually wait no, if i rethink this where every field is just a string, then the `TemplateModel.render` thing makes way more sense - so each class renders to a string, and the parent template accepts a prebaked string.

So instead of 

```jinja
{% for c in classes.values() %}

{{ class(c.name, c.bases, c.attributes, c.description, pydantic_ver) }}
{% endfor %}
```

it would be 

```python
def render(self) -> str:
    classes = [c.render() for c in self.classes]
    return self.template.render(classes=classes)
```

and just

```jinja
{% for cls in classes %}

{{ cls }}
{% endfor %}
```

and then each of the templates are just templates not macros. i like that way more. don't merge this yet i'm going to see how that feels.

that would be way way way way better for testing, modularity, and pluggability i think.

sneakers-the-rat commented:
OK check this out i gone and done it... 

Pros: 

you can do this:

```python
from linkml.generators.pydantic.template import (
    Import, Imports,
    PydanticAttribute,
    PydanticClass,
    PydanticModule,
)

imports = [
    Import(module="pydantic", objects=[{"name": "BaseModel"}, {"name": "Field"}])
]

attr = PydanticAttribute(
    name="my_field",
    annotations={"python_range": {"value": "str"}},
    title="My Field!",
    description="A Field that is mine!",
    pattern="my_.*",
)

cls = PydanticClass(
    name="MyClass",
    bases="BaseModel",
    desciption="A Class I Made!",
    attributes={"my_field": attr},
)

module = PydanticModule(imports=imports, classes={cls.name: cls})
```

And then you can do this

```python
>>> print(Imports(imports=imports).render())
from pydantic import (
    BaseModel,
    Field
)

>>> print(attr.render())
my_field: str = Field(None, title="My Field!", description="""A Field that is mine!""")

>>> print(cls.render())
class MyClass(BaseModel):
    my_field: str = Field(None, title="My Field!", description="""A Field that is mine!""")

    @validator('my_field', allow_reuse=True)
    def pattern_my_field(cls, v):
        pattern=re.compile(r"my_.*")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid my_field format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid my_field format: {v}")
        return v

>>> print(module.render())
from pydantic import (
    BaseModel,
    Field
)

metamodel_version = "None"
version = "None"

class WeakRefShimBaseModel(BaseModel):
    __slots__ = '__weakref__'


class ConfiguredBaseModel(WeakRefShimBaseModel,
                validate_assignment = True,
                validate_all = True,
                underscore_attrs_are_private = True,
                extra = "forbid",
                arbitrary_types_allowed = True,
                use_enum_values = True):
    pass


class MyClass(BaseModel):
    my_field: str = Field(None, title="My Field!", description="""A Field that is mine!""")

    @validator('my_field', allow_reuse=True)
    def pattern_my_field(cls, v):
        pattern=re.compile(r"my_.*")
        if isinstance(v,list):
            for element in v:
                if not pattern.match(element):
                    raise ValueError(f"Invalid my_field format: {element}")
        elif isinstance(v,str):
            if not pattern.match(v):
                raise ValueError(f"Invalid my_field format: {v}")
        return v


# Update forward refs
# see https://pydantic-docs.helpmanual.io/usage/postponed_annotations/
MyClass.update_forward_refs()
```

which rocks. 
- truly independent module generation with a nice interface. 
- There is only one render method that should be general across template classes, but it can also be overridden if need be. - Very testable, very transparent - one model to one template
- Clearer division of labor imo

Overriding model generation has a few methods:
- Replacing the class attribute that points to the template
- Replacing the Environment that loads the templates in the pydanticgen class
- subclassing any of the templates and hooking them into the pydanticgen serialize method (when we clean that up, which will be next PR)

and we can make other, more explicit means of hooking in as well.


Cons:
- a little more work put onto the template model classes, especially some awkward pydantic 1 and 2 adaptations.
- since the innermost model of everything is a string when it hits the template, the templates can do less introspection, but this is arguably good - the templates are very very "dumb"

anyway lmk what u think, if u like this then i'll write tests for it, otherwise we can revert back to the previous version which just had a single top-level module class that imported/used the rest of the templates as macros

sneakers-the-rat commented:
A set of pydantic models that generates pydantic models. how meta. i think this should be its own package.

anyway i'm excited to start going back through and seeing how this lets us clean up the rest of the pydantic generator :). After we merge this, then i'll finish the array reps (taking advantage of the injected imports and classes in a build result object to only import/inject when we need to rather than in every model), and then go back and clean up the mess at the end of `serialize` by linearizing the generation process.

kevinschaper commented:
This is so good! I definitely still approve. 

> A set of pydantic models that generates pydantic models. how meta. i think this should be its own package.

Having the pydantic generator produce the data model for the pydantic generator is the obvious end game. Which I say half jokingly, but I definitely have hit this same situation where I mostly just want a pydantic model, but maybe I also want or two methods on the class. It would be cool to figure out how to eventually include them in a way that doesn't just mean putting the function in a string. 

I feel like my one takeaway from pkl / javapoet was that's it was cool to see code generation exist as a standalone library, and of course I looked for whether anyone had already made a pydantic class generator as it's own thing. We might not necessarily extract it out of linkml, but I think within the project making it feel slightly independent of pydanticgen is a good step. Ultimately pydanticgen should be the bridge between scheamview and the pydantic templating classes...and maybe generators would one day even take the form of linkml transformers ( 🤯 ).

sneakers-the-rat commented:
OK! if this is good then i'll write tests for it so we can merge 

---

> I definitely have hit this same situation where I mostly just want a pydantic model, but maybe I also want or two methods on the class

now we can do that in a transparent and straightforward way :) 

- New model for `PydanticMethod` with name, signature, return type, and body
- Corresponding template
- Add `PydanticMethod` as field to `PydanticClass`

poof!

The only linkml-specific parts are the peculiar form of the `Annotation` object, but that's something we can massage out when we work on the rest of the class. 

> making it feel slightly independent of pydanticgen is a good step. 

I have been looking around the rest of the generators to see what overlaps - oocodegen has some models, but they seem to be all for javagen, so I wonder how much of this structure we can generalize to other generators? My focus is on pydanticgen bc that's the one i want to use, and all my stuff will be "plugins" on top of the sort of ad-hoc plugin system i'm making here, but would also be happy to help tidy the other ones!

> Ultimately pydanticgen should be the bridge between scheamview and the pydantic templating classes...and maybe generators would one day even take the form of linkml transformers ( 🤯 ).

Yes this is how i'm thinking about it too! truth be told I don't quite understand what the transformers do, but yes the goal is to have a declarative source and target, and then the generator is the middle part.

sneakers-the-rat commented:
OK - just wrote the docs and the tests.

Altogether for this PR, for the changelog:

Pydanticgen
- Refactor pydanticgen templates into separate modules for individual generation types
- Create TemplateModel classes that can independently render pydantic components
- Refactor pydanticgen into its own set of submodules, re-imported public classes into `__init__.py`
- Deprecate `template-file` cli and pydanticgen option. replace with `template-dir` and document and add deprecation warning
- Update format of `imports` from prior PR to reflect new `TemplateModel` system
- Make a mess at the end of `PydanticGenerator.serialize` that we will clean up in a future PR
- Tidy all pydantic templates to avoid excess whitespace
- Remove default import of `typing_extensions.Literal` import since we dropped 3.8

Docs:
- Add all members for pydanticgen to autodoc
- Add narrative docs for template usage
- enable napoleon for google style docstrings becuase regular rst docstrings are terrible
- enable intersphinx mappings and add python, pydantic, and jinja2
- Update `examples/PersonSchema/personinfo_pydantic.py` which was 2 years old and used dataclasses still!

Linting:
- Added an exception to `F401` (no unused imports) for `__init__.py` since that's a normal way to declare the public interface to a set of submodules

Tests:
- Tests for all added functionality
- Changed `https://w3id.org/linkml/types` to `linkml:types` because the former was doing network requests that were timing out in `test_pydanticgen`
- added `scripts/test_gen_pydantic.py` to test `template_dir` overrides

If tests pass, this is ready to merge :)

rly commented:
> * Added an exception to `F401` (no unused imports) for `__init__.py` since that's a normal way to declare the public interface to a set of submodules

I learned recently that [PEP8 recommends](https://peps.python.org/pep-0008/#public-and-internal-interfaces) having an `__all__` variable in `__init__.py` with a list of strings of names of the public API. If you do that, then the linters that I have used at least do not complain with F401.

sneakers-the-rat commented:
Let It Be So
